cmd/build/main.go

```diff
@@ -204,32 +82,67 @@ func run() error {
 		return fmt.Errorf("invalid image type %q for distro %q and arch %q: %w", imgTypeName, distroName, archName, err)
 	}
 
-	// get repositories
-	repos, err := loadRepos(repositories, distroName, archName)
-	if err != nil {
-		return fmt.Errorf("failed to get repositories for %s/%s: %w", distroName, archName, err)
+	var reporeg *reporegistry.RepoRegistry
+	var overrideRepos []rpmmd.RepoConfig
+	if st, err := os.Stat(repositories); err == nil && !st.IsDir() {
+		// anything that is not a dir is tried to be loaded as a file
+		// to allow "-repositories <arbitrarily-named-file>.json"
+		repoConfig, err := rpmmd.LoadRepositoriesFromFile(repositories)
+		if err != nil {
+			return fmt.Errorf("failed to load repositories from %q: %w", repositories, err)
+		}
+		overrideRepos = repoConfig[archName]
+	} else {
+		// HACK: reporegistry hardcodes adding the "repositories"
+		// dir to the "repoConfigPaths" but the interface of
+		// "images" does not expect this.
+		// XXX: should we fix this in reporegistry?
+		repositories = filepath.Join(repositories, "..")
+		reporeg, err = reporegistry.New([]string{repositories})
+		if err != nil {
+			return fmt.Errorf("failed to load repositories from %q: %w", repositories, err)
+		}
 	}
-	repos = filterRepos(repos, imgTypeName)
-	if len(repos) == 0 {
-		return fmt.Errorf("no repositories defined for %s/%s", distroName, archName)
+	seedArg, err := cmdutil.SeedArgFor(config, imgType.Name(), distribution.Name(), archName)
+	if err != nil {
+		return err
 	}
 
 	fmt.Printf("Generating manifest for %s: ", config.Name)
-	mf, err := makeManifest(config, imgType, distribution, repos, archName, rpmCacheRoot)
+	var mf bytes.Buffer
+	manifestOpts := manifestgen.Options{
+		Output:         &mf,
+		Cachedir:       filepath.Join(rpmCacheRoot, archName+distribution.Name()),
+		WarningsOutput: os.Stderr,
+		OverrideRepos:  overrideRepos,
+		CustomSeed:     &seedArg,
+	}
+	// add RHSM fact to detect changes
+	config.Options.Facts = &facts.ImageOptions{
+		APIType: facts.TEST_APITYPE,
+	}
+	if config.Blueprint == nil {
+		config.Blueprint = &blueprint.Blueprint{}
+	}
+
+	mg, err := manifestgen.New(reporeg, &manifestOpts)
 	if err != nil {
-		return err
+		return fmt.Errorf("[ERROR] manifest generator creation failed: %w", err)
+	}
+	if err := mg.Generate(config.Blueprint, distribution, imgType, arch, &config.Options); err != nil {
+		return fmt.Errorf("[ERROR] manifest generation failed: %w", err)
 	}
 	fmt.Print("DONE\n")
 
 	manifestPath := filepath.Join(buildDir, "manifest.json")
-	if err := save(mf, manifestPath); err != nil {
-		return err
+	if err := os.WriteFile(manifestPath, mf.Bytes(), 0644); err != nil {
```

Comment:
gosec doesn't like this any more it seems.  Needs 600 (or less).  That's fine here but maybe we can consider disabling that rule globally if we need to write more files with more open permissions.